substrate-node/pallets/pallet-smart-contract/src/tests.rs

```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+
+        pool_state
+            .write()
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 22);
+        run_to_block(22, Some(&mut pool_state));
+        
+        // Check if last event is Contract billed event
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        let contract_bill = types::ContractBill {
+            contract_id: node_contract_id,
+            timestamp: 1628082132,
+            discount_level,
+            amount_billed: amount_due as u128,
+        };
+        assert_eq!(
+            events
+                .last()
+                .unwrap()
+                .event,
+            MockEvent::SmartContractModule(SmartContractEvent::ContractBilled(contract_bill))
+        );
+
+        // Node contract cost settled (IP address cost), but it should remain in the grace period due to the suspended rent contract
+        // The cost of the node resources utilized by this node contract is billed to the rent contract
+        let node_contract = SmartContractModule::contracts(node_contract_id).unwrap();
+        assert_eq!(node_contract.state, types::ContractState::GracePeriod(21));
+    });
+}
+
+#[test]
+fn test_node_contract_with_overdraft_on_dedicated_node_shouldnt_be_restored_when_rent_contract_overdraft_settled() {
+    // Test ensuring that a node contract with overdraft on a dedicated node remains in the grace period when the rent contract overdraft is settled
+    let (mut ext, mut pool_state) = new_test_ext_with_pool_state(0);
+    ext.execute_with(|| {
+        run_to_block(1, Some(&mut pool_state));
+        prepare_dedicated_farm_and_node();
+
+        let node_id = 1;
+
+        // Create a rent contract
+        assert_ok!(SmartContractModule::create_rent_contract(
+            RuntimeOrigin::signed(charlie()),
+            node_id,
+            None
+        ));
+        let rent_contract_id = 1;
+
+        run_to_block(2, Some(&mut pool_state));
+
+        // Create a node contract with a public IP
+        assert_ok!(SmartContractModule::create_node_contract(
+            RuntimeOrigin::signed(charlie()),
+            node_id,
+            generate_deployment_hash(),
+            get_deployment_data(),
+            1,
+            None
+        ));
+        let node_contract_id = 2;
+        push_contract_resources_used(node_contract_id);
+
+        // Get spendable balance
+        let spendable_balance = SmartContractModule::get_usable_balance(&charlie());
+
+        // The grace period for the rent contract should start
+        // Node contract should also transition to grace period
+        pool_state
+            .write()
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 11);
+        run_to_block(11, Some(&mut pool_state));
+
+        pool_state
+            .write()
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 12);
+        run_to_block(12, Some(&mut pool_state));
+
```

Comment:
also here something like:
~~~rust
assert_ne!(node_contract_overdraft, 0);
assert_eq!(node_contract_reserve, 0);
assert_eq!(rent_contract_overdraft, node_contract_ip_cost_1_cycle);
~~~

to be able to understand better what is happening in "backstage"